tool/tbot/botfs/fs_linux.go

```diff
@@ -0,0 +1,454 @@
+		// of the world, though: if an attacker attempts a symlink attack we'll
+		// just open the correct file for read/write later (and error when it
+		// doesn't exist).
+		if err := os.Mkdir(path, DefaultDirMode); err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		f, err := openSecure(path)
+		if err == unix.ENOSYS {
+			// bubble up the original error for comparison
+			return err
+		} else if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// No writing to do, just close it.
+		f.Close()
+	}
+
+	return nil
+}
+
+// Create attempts to create the given file or directory with the given
+// symlinks mode.
+func Create(path string, isDir bool, symlinksMode SymlinksMode) error {
+	// Implementation note: paranoid file _creation_ is only really useful for
+	// providing an early warning if openat2() / ACLs are unsupported on the
+	// host system, as it will catch compatibility issues during `tbot init`.
+	// Read() and Write() with Symlinks(Try)Secure are the codepaths that
+	// actually prevents symlink attacks.
+
+	switch symlinksMode {
+	case SymlinksSecure:
+		if err := createSecure(path, isDir); err != nil {
+			if err == unix.ENOSYS {
+				return trace.Errorf("createSecure(%q) failed due to missing "+
+					"syscall; `symlinks: insecure` may be required for this "+
+					"system", path)
+			}
+
+			return trace.Wrap(err)
+		}
+	case SymlinksTrySecure:
+		err := createSecure(path, isDir)
+		if err == nil {
+			// All good, move on.
+			return nil
+		}
+
+		if err != unix.ENOSYS {
+			// Something else went wrong, fail.
+			return trace.Wrap(err)
+		}
+
+		// TODO: this will be very noisy on older systems. Maybe flip a global
+		// or something to only log once?
+		log.Warnf("Failed to create %q securely due to missing syscall; "+
+			"falling back to regular file creation. Set `symlinks: "+
+			"insecure` on this destination to disable this warning.", path)
+
+		return trace.Wrap(createStandard(path, isDir))
+	case SymlinksInsecure:
+		return trace.Wrap(createStandard(path, isDir))
+	}
+
+	return nil
+}
+
+// Read reads the contents of the given file into memory.
+func Read(path string, symlinksMode SymlinksMode) ([]byte, error) {
+	file, err := openSymlinksMode(path, symlinksMode)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	defer file.Close()
+
+	// Blatantly stolen from go's os/file.go since there's no stdlib function
+	// for reading everything from an _already open_ file, sigh.
+	var size int
+	if info, err := file.Stat(); err == nil {
+		size64 := info.Size()
+		if int64(int(size64)) == size64 {
+			size = int(size64)
+		}
+	}
+	size++ // one byte for final read at EOF
+
+	// If a file claims a small size, read at least 512 bytes.
+	// In particular, files in Linux's /proc claim size 0 but
+	// then do not work right if read in small pieces,
+	// so an initial read of 1 byte would not work correctly.
+	if size < 512 {
+		size = 512
+	}
+
+	data := make([]byte, 0, size)
+	for {
+		if len(data) >= cap(data) {
```

Comment:
This is a bit confusing, as Go won't allow `len > cap`. What are you actually trying to accomplish here?